File: src/encode_scripts/rsa.py
# Name: RSA 加密
# Description: 非对称加密 (生成密钥、公钥加密、私钥解密)
# Avatar: 🔐

from Crypto.PublicKey import RSA
from Crypto.Cipher import PKCS1_OAEP
import base64
import json

# Global config placeholder (injected by JS)
SCRIPT_CONFIG = {}

def get_config_schema():
    """
    Define the configuration UI for this script.
    """
    return json.dumps({
        "params": [
            {
                "name": "public_key",
                "label": "公钥 (Public Key) - 用于加密",
                "type": "textarea",
                "placeholder": "-----BEGIN PUBLIC KEY-----..."
            },
            {
                "name": "private_key",
                "label": "私钥 (Private Key) - 用于解密",
                "type": "textarea",
                "placeholder": "-----BEGIN RSA PRIVATE KEY-----..."
            }
        ],
        "actions": [
            {
                "name": "generate_keys",
                "label": "生成新密钥对 (Generate New Keys)",
                "type": "button"
            }
        ]
    })

def generate_keys():
    """
    Generate a new RSA key pair and return them to update the config.
    """
    key = RSA.generate(2048)
    private_key = key.export_key().decode('utf-8')
    public_key = key.publickey().export_key().decode('utf-8')
    return json.dumps({
        "public_key": public_key,
        "private_key": private_key
    })

def encode(text):
    """
    使用配置中的公钥进行加密
    """
    if not text:
        return ''
        
    pub_key_pem = SCRIPT_CONFIG.get('public_key', '').strip()
    if not pub_key_pem:
        raise ValueError("请先在设置中配置公钥 (Public Key)")

    try:
        recipient_key = RSA.import_key(pub_key_pem)
        cipher_rsa = PKCS1_OAEP.new(recipient_key)
        encrypted_msg = cipher_rsa.encrypt(text.encode('utf-8'))
        return base64.b64encode(encrypted_msg).decode('utf-8')
    except ValueError as ve:
         raise ValueError(f"公钥格式错误: {str(ve)}")
    except Exception as e:
         raise ValueError(f"加密错误: {str(e)}")

def decode(text):
    """
    使用配置中的私钥进行解密
    """
    if not text:
        return ''
        
    priv_key_pem = SCRIPT_CONFIG.get('private_key', '').strip()
    if not priv_key_pem:
        raise ValueError("请先在设置中配置私钥 (Private Key)")
        
    try:
        private_key = RSA.import_key(priv_key_pem)
        cipher_rsa = PKCS1_OAEP.new(private_key)
        
        # Decode base64 first
        try:
            ciphertext = base64.b64decode(text)
        except:
             raise ValueError("输入不是有效的Base64编码")
             
        decrypted_msg = cipher_rsa.decrypt(ciphertext)
        return decrypted_msg.decode('utf-8')
    except ValueError as ve:
         raise ValueError(f"密钥或密文错误: {str(ve)}")
    except Exception as e:
        raise ValueError(f"解密错误: {str(e)} (请检查私钥是否匹配)")

File: src/encode_scripts/rsa.ts
// Name: RSA 加密
// Description: 非对称加密 (生成密钥、公钥加密、私钥解密)
// Avatar: 🔑

function get_config_schema() {
    return {
        params: [
            {
                name: "public_key",
                label: "公钥 (Public Key) - 用于加密",
                type: "textarea",
                placeholder: "-----BEGIN PUBLIC KEY-----..."
            },
            {
                name: "private_key",
                label: "私钥 (Private Key) - 用于解密",
                type: "textarea",
                placeholder: "-----BEGIN PRIVATE KEY-----..."
            }
        ],
        actions: [
            {
                name: "generate_keys",
                label: "生成新密钥对 (Generate New Keys)",
                type: "button"
            }
        ]
    };
}

function pemToArrayBuffer(pem) {
    const b64 = pem.replace(/-----[A-Z ]+-----/g, '').replace(/\s/g, '');
    const binary = atob(b64);
    const bytes = new Uint8Array(binary.length);
    for (let i = 0; i < binary.length; i++) {
        bytes[i] = binary.charCodeAt(i);
    }
    return bytes.buffer;
}

function arrayBufferToPem(buffer, type) {
    const bytes = new Uint8Array(buffer);
    let binary = '';
    for (let i = 0; i < bytes.length; i++) {
        binary += String.fromCharCode(bytes[i]);
    }
    const b64 = btoa(binary);
    const lines = b64.match(/.{1,64}/g).join('\n');
    return '-----BEGIN ' + type + '-----\n' + lines + '\n-----END ' + type + '-----';
}

async function generate_keys() {
    const keyPair = await crypto.subtle.generateKey(
        {
            name: 'RSA-OAEP',
            modulusLength: 2048,
            publicExponent: new Uint8Array([1, 0, 1]),
            hash: 'SHA-256'
        },
        true,
        ['encrypt', 'decrypt']
    );
    const publicKey = await crypto.subtle.exportKey('spki', keyPair.publicKey);
    const privateKey = await crypto.subtle.exportKey('pkcs8', keyPair.privateKey);
    return JSON.stringify({
        public_key: arrayBufferToPem(publicKey, 'PUBLIC KEY'),
        private_key: arrayBufferToPem(privateKey, 'PRIVATE KEY')
    });
}

async function encode(text) {
    if (!text) return '';

    const pubKeyPem = (SCRIPT_CONFIG.public_key || '').trim();
    if (!pubKeyPem) {
        throw new Error("请先在设置中配置公钥 (Public Key)");
    }

    try {
        const keyData = pemToArrayBuffer(pubKeyPem);
        const key = await crypto.subtle.importKey(
            'spki', keyData,
            { name: 'RSA-OAEP', hash: 'SHA-256' },
            false, ['encrypt']
        );
        const encrypted = await crypto.subtle.encrypt(
            { name: 'RSA-OAEP' },
            key,
            new TextEncoder().encode(text)
        );
        const bytes = new Uint8Array(encrypted);
        let binary = '';
        for (let i = 0; i < bytes.length; i++) {
            binary += String.fromCharCode(bytes[i]);
        }
        return btoa(binary);
    } catch (e) {
        const msg = e.message || String(e);
        if (msg.indexOf("请先") !== -1) throw e;
        throw new Error("加密错误: " + msg);
    }
}

async function decode(text) {
    if (!text) return '';

    const privKeyPem = (SCRIPT_CONFIG.private_key || '').trim();
    if (!privKeyPem) {
        throw new Error("请先在设置中配置私钥 (Private Key)");
    }

    try {
        let binary;
        try {
            binary = atob(text);
        } catch (e) {
            throw new Error("输入不是有效的Base64编码");
        }

        const ciphertext = new Uint8Array(binary.length);
        for (let i = 0; i < binary.length; i++) {
            ciphertext[i] = binary.charCodeAt(i);
        }

        const keyData = pemToArrayBuffer(privKeyPem);
        const key = await crypto.subtle.importKey(
            'pkcs8', keyData,
            { name: 'RSA-OAEP', hash: 'SHA-256' },
            false, ['decrypt']
        );
        const decrypted = await crypto.subtle.decrypt(
            { name: 'RSA-OAEP' },
            key,
            ciphertext
        );
        return new TextDecoder().decode(decrypted);
    } catch (e) {
        const msg = e.message || String(e);
        if (msg.indexOf("请先") !== -1 || msg.indexOf("输入不是") !== -1) throw e;
        throw new Error("解密错误: " + msg + " (请检查私钥是否匹配)");
    }
}
